clustering.py

```diff
@@ -244,3 +246,220 @@ def sort_labels(label_dict, cluster_labels, ts_to_labels):
+
+def update_centroids(data, clusters, centroids):
+    """Updates the centroids based on the elements in each cluster.
+
+    Args:
+        data: An np array where each row is a time series and each
+            column is a time.
+        clusters: A list where the ith element is a list of the indexes
+            of the elements in the ith cluster.
+        centroids: A list of the centroids.
+
+    Returns:
+        True if the centroids were updated without an error, False
+        otherwise.
+    """
+    valid_clusters = True
+    for cluster_ind, _ in enumerate(centroids):
+        if len(clusters[cluster_ind]) == 0:
+            valid_clusters = False
+            break
+        if valid_clusters:
+            total = np.zeros((centroids.shape[1]))
+            for ts_index in clusters[cluster_ind]:
+                total += data[ts_index]
+            centroids[cluster_ind] = total / len(clusters[cluster_ind])
+    return valid_clusters
+
+def k_means_init(data, num_clusters, run_num):
+    """Runs k-means++ initialization which aims to spread out the
+    cluster centroids.
+
+    Args:
+        data: An np array where each row is a time series and each
+            column is a time.
+        num_clusters: The number of clusters.
+        run_num: The number of times k-means has been reinitialized. It
+            is used as the seed for np.random.seed.
+
+    Returns:
+        An np array where the ith element is the ith cluster centroid.
+    """
+    np.random.seed(run_num)
+    num_ts = len(data)
+    first = int(num_ts * .2)
+    centroids = np.array([data[first]])
+
+    for _ in range(num_clusters -1):
+        _, distances = pairwise_distances_argmin_min(data, centroids)
+        choices = np.random.choice(num_ts, 1, p=distances/np.sum(distances))
+        picked = choices[0]
+        centroids = np.append(centroids, [data[picked]], axis=0)
+    return centroids
+
+def violates_cons(option, ts_index, ts_to_cluster, must_link, can_not_link):
+    """Checks if any constraint is violated by placing ts_index in the
+    cluster option.
+
+    Args:
+        option: The index of the cluster where the time series ts_index
+            may be placed.
+        ts_index: The index of the time series.
+        ts_to_cluster: A dictionary where each key is a time series
+            index and each value is the index of the cluster the time
+            series is assigned to.
+        must_link: A dictionary where each key is a time series index
+            and each value is a list of time series indexes which must
+            be in the same cluster.
+        can_not_link: A dictionary where each key is a time series index
+            and each value is a list of time series indexes which can
+            not be in the same cluster as the key.
+
+    Returns:
+        True if a constraint was violated and False otherwise.
+    """
+    if ts_index in must_link:
+        for ts_2 in must_link[ts_index]:
```

Comment:
Why is this called ts_2?

Reply:
ts_1 is the first time series with index "ts_index" and ts_2 refers to the second time series with index "ts_2" which is confusing. I can update that.